### PR TITLE
[llvm][AArch64][Assembly]: Add SME_F8F16 and SME_F8F32 Ass/Disass.

### PR DESCRIPTION
**Pull request by hassnaaHamdi**
This patch adds the feature flags of SME_F8F16 and SME_F8F32,
 and the assembly/disassembly for the following instructions of SME2:

  * SME:
    - FMLAL, FMLALL
    - FVDOT, FVDOTT
    - FVDOTB 
    - FMOPA
   
That is according to this documentation:
https://developer.arm.com/documentation/ddi0602/2023-09

### CODE REVIEW

File: llvm/lib/Target/AArch64/AsmParser/AArch64AsmParser.cpp [hunk 1]
```diff
@@ -4546,18 +4548,21 @@ ParseStatus AArch64AsmParser::tryParseZTOperand(OperandVector &Operands) {
     if (getParser().parseExpression(ImmVal))
       return ParseStatus::NoMatch;
     const MCConstantExpr *MCE = dyn_cast<MCConstantExpr>(ImmVal);
+    Operands.push_back(AArch64Operand::CreateImm(
+        MCConstantExpr::create(MCE->getValue(), getContext()), StartLoc,
+        getLoc(), getContext()));
     if (!MCE)
```
Comment (CarolineConcatto) [inline llvm/lib/Target/AArch64/AsmParser/AArch64AsmParser.cpp:4554]: I think we should change the order :
```
if (!MCE)
  return TokError("immediate value expected for vector index");

    Operands.push_back(AArch64Operand::CreateImm(
        MCConstantExpr::create(MCE->getValue(), getContext()), StartLoc,
        getLoc(), getContext()));

```
Just because it is only acceptable to get the value (MCE->getValue()) and push to the operand if we could actually create the constant.

File: llvm/lib/Target/AArch64/AsmParser/AArch64AsmParser.cpp [hunk 2]
```diff
@@ -4546,18 +4548,21 @@ ParseStatus AArch64AsmParser::tryParseZTOperand(OperandVector &Operands) {
...
       return TokError("immediate value expected for vector index");
+    if (getTok().is(AsmToken::Comma)) {
```
Comment (CarolineConcatto) [inline llvm/lib/Target/AArch64/AsmParser/AArch64AsmParser.cpp:4556]: nit: I believe we can use:
   if (parseOptionalToken(AsmToken::Comma))
and then remove Lex(); and the curly brackets
Something like 
```
   if (parseOptionalToken(AsmToken::Comma))
      if (parseOptionalMulOperand(Operands))
           return MatchOperand_ParseFail;
```

File: llvm/lib/Target/AArch64/SMEInstrFormats.td [hunk 3]
```diff
@@ -1937,7 +1948,6 @@ class sme2_mla_long_array_vg24_single<bits<2> op0, bit vg4, bits<2> op, bit o2,
   let Inst{1-0}   = imm;
 }
 
-
```
Comment (CarolineConcatto) [inline llvm/lib/Target/AArch64/SMEInstrFormats.td:1940]: Unrelated change

File: llvm/lib/Target/AArch64/SMEInstrFormats.td [hunk 4]
```diff
@@ -2014,7 +2025,6 @@ class sme2_mla_long_array_vg2_multi<string mnemonic, bits<2> op0, bits<3> op,
 multiclass sme2_fp_mla_long_array_vg2_multi<string mnemonic, bits<3> op, MatrixOperand matrix_ty,
                                             RegisterOperand multi_vector_ty,
                                             ValueType zpr_ty, SDPatternOperator intrinsic> {
-
```
Comment (CarolineConcatto) [inline llvm/lib/Target/AArch64/SMEInstrFormats.td:2017]: Unrelated change

File: llvm/lib/Target/AArch64/SMEInstrFormats.td [hunk 5]
```diff
@@ -4713,3 +4830,36 @@ class sme2p1_luti4_vector_vg4_index<bits<2> sz, RegisterOperand vector_ty,
 multiclass sme2p1_luti4_vector_vg4_index<string mnemonic> {
   def _H: sme2p1_luti4_vector_vg4_index<0b01, ZZZZ_h_strided, VectorIndexD, mnemonic>;
 }
+
+// SME2 lookup table two source registers expand to four contiguous destination registers
```
Comment (CarolineConcatto) [inline llvm/lib/Target/AArch64/SMEInstrFormats.td:4834]: I believe we should not have this class here.

File: llvm/lib/Target/AArch64/SMEInstrFormats.td [hunk 6]
```diff
@@ -4713,3 +4830,36 @@ class sme2p1_luti4_vector_vg4_index<bits<2> sz, RegisterOperand vector_ty,
...
+  let Inst{13-12} = sz;
+  let Inst{11-10} = op;
+  let Inst{9-6} = Zn;
+  let Inst{5}   = 0b0;
+  let Inst{4-2} = Zd;
+  let Inst{1-0} = 0b00;
+}
+
+// SME2 lookup table two source registers expand to four non-contiguous destination registers
+class sme2_luti4_vector_vg4_strided<bits<2> sz, bits<2> op, string mnemonic>
```
Comment (CarolineConcatto) [inline llvm/lib/Target/AArch64/SMEInstrFormats.td:4851]: This class should be removed.

File: llvm/lib/Target/AArch64/SMEInstrFormats.td [hunk 7]
```diff
@@ -2608,7 +2630,83 @@ multiclass sme2_multi_vec_array_vg4_index_64b<string mnemonic, bits<3> op,
...
 }
+
+// FMLAL (multiple and indexed vector, FP8 to FP16)
+class sme2_multi_vec_array_vg24_index_16b<bits<2> sz, bit vg4, bits<3> op,
+                                          RegisterOperand multi_vector_ty, string mnemonic>
+    : I<(outs MatrixOp16:$ZAda),
+        (ins MatrixOp16:$_ZAda, MatrixIndexGPR32Op8_11:$Rv, uimm2s2range:$imm2,
+         multi_vector_ty:$Zn, ZPR4b8:$Zm, VectorIndexB:$i),
+         mnemonic, "\t$ZAda[$Rv, $imm2, " # !if(vg4, "vgx4", "vgx2") # "], $Zn, $Zm$i",
+        "", []>, Sched<[]> {
```
Comment (CarolineConcatto) [inline llvm/lib/Target/AArch64/SMEInstrFormats.td:2641]: nit:align

File: llvm/lib/Target/AArch64/SMEInstrFormats.td [hunk 8]
```diff
@@ -2448,11 +2457,24 @@ multiclass sme2p1_multi_vec_array_vg2_index_16b<string mnemonic, bits<2> sz, bit
...
   def : InstAlias<mnemonic # "\t$ZAda[$Rv, $imm3], $Zn, $Zm$i",
         (!cast<Instruction>(NAME) MatrixOp16:$ZAda,  MatrixIndexGPR32Op8_11:$Rv, sme_elm_idx0_7:$imm3,
         multi_vector_ty:$Zn, zpr_ty:$Zm, VectorIndexH:$i), 0>;
 }
 
+// SME2 multi-vec indexed FP8 two-way vertical dot product to single precision
+// two registers
+class sme2_fp8_multi_vec_array_vg4_index<string mnemonic, bit T>
+   : sme2_multi_vec_array_vg2_index<0b11, {0b01,?,0b0, T,?}, MatrixOp32,
+                                     ZZ_b_mul_r, ZPR4b8, VectorIndexS, mnemonic> {
```
Comment (CarolineConcatto) [inline llvm/lib/Target/AArch64/SMEInstrFormats.td:2470]: nit:ling

**Review by CarolineConcatto — CHANGES_REQUESTED** [review]
Thank  you Hassnaa for this patch. I have some comments left.

File: llvm/lib/Target/AArch64/SMEInstrFormats.td [hunk 9]
```diff
@@ -3059,6 +3158,25 @@ class sme2_movt_scalar_to_zt<string mnemonic, bits<7> opc>
   let Inst{4-0}   = Rt;
 }
 
+// SME2 move vector to lookup table
+class sme2_movt_zt_to_zt<string mnemonic, bits<7> opc>
```
Comment (CarolineConcatto) [inline llvm/lib/Target/AArch64/SMEInstrFormats.td:3162]: Why this class is here? It is not used in any other place.

File: llvm/lib/Target/AArch64/AArch64SMEInstrInfo.td [hunk 10]
```diff
@@ -885,3 +885,57 @@ defm FAMIN_2Z2Z : sme2_fp_sve_destructive_vector_vg2_multi<"famin", 0b0010101>;
...
+def  FMLAL_VG2_MZZ_BtoH   : sme2_mla_long_array_single_16b<"fmlal">;
+// TODO: Replace nxv16i8 by nxv16f8
+defm FMLAL_VG2_M2ZZ_BtoH  : sme2_fp_mla_long_array_vg2_single<"fmlal",  0b001, MatrixOp16, ZZ_b, ZPR4b8, nxv16i8, null_frag>;
+defm FMLAL_VG4_M4ZZ_BtoH  :  sme2_fp_mla_long_array_vg4_single<"fmlal", 0b001, MatrixOp16, ZZZZ_b, ZPR4b8, nxv16i8, null_frag>;
+defm FMLAL_VG2_M2Z2Z_BtoH : sme2_fp_mla_long_array_vg2_multi<"fmlal",   0b100, MatrixOp16, ZZ_b_mul_r, nxv16i8, null_frag>;
+defm FMLAL_VG4_M4Z4Z_BtoH : sme2_fp_mla_long_array_vg4_multi<"fmlal",   0b100, MatrixOp16, ZZZZ_b_mul_r, nxv16i8, null_frag>;
+
+defm FMOPA_MPPZZ_BtoH : sme2p1_fmop_tile_fp16<"fmopa", 0b1, 0b0, 0b01, ZPR8>;
+
+} //[HasSME2p1, HasSMEF8F16]
```
Comment (CarolineConcatto) [inline llvm/lib/Target/AArch64/AArch64SMEInstrInfo.td:912]: This is HasSME2

File: llvm/lib/Target/AArch64/AArch64SMEInstrInfo.td [hunk 11]
```diff
@@ -885,3 +885,57 @@ defm FAMIN_2Z2Z : sme2_fp_sve_destructive_vector_vg2_multi<"famin", 0b0010101>;
...
+defm FMLALL_MZZ_BtoS       : sme2_mla_ll_array_single<"fmlall",      0b01000, MatrixOp32, ZPR8, ZPR4b8, nxv16i8, null_frag>;
+defm FMLALL_VG2_M2ZZ_BtoS  : sme2_mla_ll_array_vg24_single<"fmlall", 0b000001, MatrixOp32, ZZ_b, ZPR4b8>;
+defm FMLALL_VG4_M4ZZ_BtoS  : sme2_mla_ll_array_vg24_single<"fmlall", 0b010001, MatrixOp32, ZZZZ_b, ZPR4b8>;
+defm FMLALL_VG2_M2Z2Z_BtoS : sme2_mla_ll_array_vg2_multi<"fmlall",   0b01000, MatrixOp32, ZZ_b_mul_r, nxv16i8, null_frag>;
+defm FMLALL_VG4_M4Z4Z_BtoS : sme2_mla_ll_array_vg4_multi<"fmlall",   0b01000, MatrixOp32, ZZZZ_b_mul_r, nxv16i8, null_frag>;
+
+
+defm FMOPA_MPPZZ_BtoS : sme_outer_product_fp32<0b0, 0b01, ZPR8, "fmopa", null_frag>;
+
+} //[HasSME2p1, HasSMEF8F32]
```
Comment (CarolineConcatto) [inline llvm/lib/Target/AArch64/AArch64SMEInstrInfo.td:940]: This is HasSME2

File: llvm/test/MC/AArch64/FP8_SME2/mopa-diagnostics.s [hunk 12]
```diff
@@ -0,0 +1,46 @@
+// RUN: not llvm-mc -triple=aarch64 -show-encoding -mattr=+sme2p1,+sme-lutv2 2>&1 < %s | FileCheck %s
```
Comment (CarolineConcatto) [inline llvm/test/MC/AArch64/FP8_SME2/mopa-diagnostics.s:1]: Is this correct? Should it not be :  -mattr=+sme2,+sme-f8f16,+sme-f8f32

File: llvm/test/MC/AArch64/FP8_SME2/mla-diagnostics.s [hunk 13]
```diff
@@ -0,0 +1,195 @@
+// RUN: not llvm-mc -triple=aarch64 -show-encoding -mattr=+sme2p1,+sme-f8f16,+sme-f8f32  2>&1 < %s | FileCheck %s
```
Comment (CarolineConcatto) [inline llvm/test/MC/AArch64/FP8_SME2/mla-diagnostics.s:1]: We can replace sme2p1, by sme2

File: llvm/test/MC/AArch64/FP8_SME2/dot-diagnostics.s [hunk 14]
```diff
@@ -0,0 +1,215 @@
+// RUN: not llvm-mc -triple=aarch64 -show-encoding -mattr=+sme2p1,+sme-f8f16,+sme-f8f32  2>&1 < %s | FileCheck %s
```
Comment (CarolineConcatto) [inline llvm/test/MC/AArch64/FP8_SME2/dot-diagnostics.s:1]: Same here can you replace sme2p1 by sme2

File: llvm/lib/Target/AArch64/AArch64SMEInstrInfo.td [hunk 15]
```diff
@@ -885,3 +885,57 @@ defm FAMIN_2Z2Z : sme2_fp_sve_destructive_vector_vg2_multi<"famin", 0b0010101>;
...
+
+defm FDOT_VG2_M2ZZI_BtoH : sme2p1_multi_vec_array_vg2_index_16b<"fdot",    0b11, 0b010, ZZ_b_mul_r, ZPR4b8>;
+defm FDOT_VG4_M4ZZI_BtoH : sme2p1_multi_vec_array_vg4_index_16b<"fdot",    0b100, ZZZZ_b_mul_r, ZPR4b8>;
+defm FDOT_VG2_M2ZZ_BtoH  :  sme2_dot_mla_add_sub_array_vg24_single<"fdot", 0b0010001, MatrixOp16, ZZ_b, ZPR4b8>;
+defm FDOT_VG4_M4ZZ_BtoH  :  sme2_dot_mla_add_sub_array_vg24_single<"fdot", 0b0110001, MatrixOp16, ZZZZ_b, ZPR4b8>;
+// TODO: Replace nxv16i8 by nxv16f8
+defm FDOT_VG2_M2Z2Z_BtoH : sme2_dot_mla_add_sub_array_vg2_multi<"fdot",    0b0100100, MatrixOp16, ZZ_b_mul_r, nxv16i8, null_frag>;
+defm FDOT_VG4_M4Z4Z_BtoH : sme2_dot_mla_add_sub_array_vg4_multi<"fdot",    0b0100100, MatrixOp16, ZZZZ_b_mul_r, nxv16i8, null_frag>;
+
+def  FMLAL_MZZI_BtoH      : sme2_mla_ll_array_index_16b<"fmlal", 0b11, 0b00>;
```
Comment (CarolineConcatto) [inline llvm/lib/Target/AArch64/AArch64SMEInstrInfo.td:900]: Align

File: llvm/include/llvm/TargetParser/AArch64TargetParser.h [hunk 16]
```diff
@@ -285,6 +287,8 @@ inline constexpr ExtensionInfo Extensions[] = {
     {"ssve-fp8dot2", AArch64::AEK_SSVE_FP8DOT2, "+ssve-fp8dot2", "-ssve-fp8dot2", FEAT_INIT, "+sme2", 0},
     {"fp8dot4", AArch64::AEK_FP8DOT4, "+fp8dot4", "-fp8dot4", FEAT_INIT, "", 0},
     {"ssve-fp8dot4", AArch64::AEK_SSVE_FP8DOT4, "+ssve-fp8dot4", "-ssve-fp8dot4", FEAT_INIT, "+sme2", 0},
+    {"sme-f8f16", AArch64::AEK_SMEF8F16, "+sme-f8f16", "-sme-f8f16", FEAT_INIT, "+sme2,+fp8", 0},
```
[the remark this answers is not in the document]
Comment (hassnaaHamdi) [reply llvm/include/llvm/TargetParser/AArch64TargetParser.h:290]: why is it different for this only ? I mean as long as the others exist, so why not we keep this ?

File: llvm/lib/Target/AArch64/AArch64.td [hunk 17]
```diff
@@ -535,6 +535,12 @@ def FeatureFP8DOT4: SubtargetFeature<"fp8dot4", "HasFP8DOT4", "true",
 def FeatureSSVE_FP8DOT4 : SubtargetFeature<"ssve-fp8dot4", "HasSSVE_FP8DOT4", "true",
   "Enable SVE2 fp8 4-way dot product instructions (FEAT_SSVE_FP8DOT4)", [FeatureSME2]>;
 
+def FeatureSMEF8F16 : SubtargetFeature<"sme-f8f16", "HasSMEF8F16", "true",
+  "Enable Scalable Matrix Extension (SME) F8F16 instructions(FEAT_SME_F8F16)", [FeatureSME2, FeatureFP8]>;
```
[the remark this answers is not in the document]
Comment (hassnaaHamdi) [reply llvm/lib/Target/AArch64/AArch64.td:539]: Doesn't the "HasSMEF8F16" depend on this feature definition ?
And HasSMEFP8FP16 is required by some instances like FMOPA_MPPZZ_BtoH for example.

File: llvm/test/MC/AArch64/FP8_SME2/dot-diagnostics.s [hunk 18]
```diff
@@ -0,0 +1,215 @@
+// RUN: not llvm-mc -triple=aarch64 -show-encoding -mattr=+sme2,+sme-f8f16,+sme-f8f32  2>&1 < %s | FileCheck %s
```
Comment (CarolineConcatto) [inline llvm/test/MC/AArch64/FP8_SME2/dot-diagnostics.s:1]: Can remove +sme2

**Review by CarolineConcatto — APPROVED** [review]
Thank you Hassnaa for the work.
Before merging, remove +sme2 from the diagnostic tests.
